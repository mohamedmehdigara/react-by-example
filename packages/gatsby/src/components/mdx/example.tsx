import React, { FC, ReactNode, useCallback, useState } from 'react'
import styled from 'styled-components'
import { CodeSandboxIcon, ResetIcon } from '../icons'

const Render = styled.div`
  display: flex;
  padding: 1em;
  border: 1px solid var(--app-foreground);
  border-radius: var(--app-border-radius);
  align-items: center;
  justify-content: center;
`
const FlexRight = styled.div`
  display: flex;
  justify-content: flex-end;
`

const CodeSandboxLink = styled.a`
  display: flex;
`

const ResetButton = styled.button`
  display: flex;
  margin: 0;
  padding: 0;
  color: var(--app-foreground);
  background: none;
  border: none;
  cursor: pointer;

  :focus,
  :active {
    color: var(--app-cyan);
    outline: 2px solid var(--app-cyan);
  }
`

export interface ExampleProps {
  code?: ReactNode
  render?: ReactNode
  link?: string
}

export const Example: FC<ExampleProps> = ({ code, render, link }) => {
  const [resetKey, setResetKey] = useState(Date.now())
  const triggerReset = useCallback(() => {
    setResetKey(Date.now())
  }, [])
  return (
    <>
      {code}
      <Render key={resetKey}>
        <div>{render}</div>
      </Render>
      {link && (
        <FlexRight>
          <ResetButton
            type="button"
            title="reset example"
            onClick={triggerReset}
          >
            <ResetIcon />
          </ResetButton>
          <CodeSandboxLink
            target="_blank"
            rel="noreferrer"
            href={link}
            title="edit on CodeSandbox"
          >
            <CodeSandboxIcon />
          </CodeSandboxLink>
        </FlexRight>
      )}
    </>
  )
}
