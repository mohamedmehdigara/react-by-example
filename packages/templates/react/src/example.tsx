import React, { FC, ReactElement } from 'react'

export const Simple: FC = () => <h1>Hello world!</h1>

export default (): ReactElement => <Simple />
