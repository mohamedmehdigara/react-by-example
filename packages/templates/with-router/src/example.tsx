import React, { FC } from 'react'

export const Simple: FC = () => <h1>Hello world!</h1>

export default <Simple />
