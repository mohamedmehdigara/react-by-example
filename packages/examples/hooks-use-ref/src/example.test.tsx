import {
  cleanup,
  fireEvent,
  render,
  RenderResult,
} from '@testing-library/react'
import React from 'react'
import { PlayPause } from './example'

describe('PlayPause', () => {
  let component: RenderResult
  let video: HTMLVideoElement
  let playSpy: jest.SpyInstance
  let pauseSpy: jest.SpyInstance
  let button: HTMLElement

  beforeEach(() => {
    component = render(<PlayPause />)
    video = component.getByLabelText('video') as HTMLVideoElement

    let paused = true
    jest.spyOn(video, 'paused', 'get').mockImplementation(() => paused)
    if (video.autoplay) {
      paused = false
    }
    playSpy = jest.spyOn(video, 'play').mockImplementation(() => {
      paused = false
      return Promise.resolve()
    })
    pauseSpy = jest.spyOn(video, 'pause').mockImplementation(() => {
      paused = true
    })

    button = component.getByRole('button')
  })

  afterEach(() => {
    cleanup()
    jest.restoreAllMocks()
  })

  it('should snapshot', () => {
    expect(component.container).toMatchSnapshot()
  })

  it('should start playing', () => {
    expect(video.paused).toBe(false)
  })

  describe('when pressing the button', () => {
    beforeEach(() => {
      fireEvent.click(button)
    })

    it('should pause', () => {
      expect(pauseSpy).toHaveBeenCalledWith()
      expect(pauseSpy).toHaveBeenCalledTimes(1)
      expect(video.paused).toBe(true)
    })

    describe('when pressing the button', () => {
      beforeEach(() => {
        fireEvent.click(button)
      })

      it('should play', () => {
        expect(playSpy).toHaveBeenCalledWith()
        expect(playSpy).toHaveBeenCalledTimes(1)
        expect(video.paused).toBe(false)
      })

      describe('when pressing the button', () => {
        beforeEach(() => {
          fireEvent.click(button)
        })

        it('should pause', () => {
          expect(pauseSpy).toHaveBeenCalledWith()
          expect(pauseSpy).toHaveBeenCalledTimes(2)
          expect(video.paused).toBe(true)
        })
      })
    })
  })
})
