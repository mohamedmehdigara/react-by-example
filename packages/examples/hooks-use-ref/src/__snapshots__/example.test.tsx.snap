// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PlayPause should snapshot 1`] = `
<div>
  <video
    aria-label="video"
    autoplay=""
    loop=""
    src="https://media.tenor.com/videos/632c96bbc411d8baa3f7f43692474808/webm"
  />
  <br />
  <button
    type="button"
  >
    ‚èØ
  </button>
</div>
`;
